web/src/LanguageSelector.js

```diff
@@ -11,49 +11,74 @@ import {
   ModalVariant
 } from "@patternfly/react-core"
 
-export default function LanguageSelector({ value, onChange = () => {} }) {
-  const [isFormOpen, setFormOpen] = useState(false);
-  const [language, setLanguage] = useState(value);
-  const [options, setOptions] = useState([]);
+const reducer = (state, action) => {
+  switch (action.type) {
+    case "LOAD": {
+      return { ...state, ...action.payload };
+    }
+    case "ACCEPT": {
+      return { ...state, isFormOpen: false };
+    }
+
+    case "CANCEL": {
+      return { ...state, isFormOpen: false, current: state.initial };
+    }
+
+    case "CHANGE": {
+      return { ...state, current: action.payload };
+    }
+
+    case "OPEN": {
+      return { ...state, isFormOpen: true };
+    }
+
+    default: {
+        return state;
+    }
+  }
+}
+
+const initialState = {
+  languages: [], initial: null, current: null, isFormOpen: false
+};
+
+export default function LanguageSelector() {
   const client = useInstallerClient();
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const { current: language, languages, isFormOpen } = state;
 
-  useEffect(() => {
-    client.getLanguages().then(setOptions);
+  useEffect(async () => {
+    const languages = await client.getLanguages();
```

Comment:
Note for the future: to improve the UI feedback, maybe we can add a `loading` property in the state and put it as true here. That way, we can put a spinner somewhere to give feedback to the user while waiting for promises.

Reply:
Yes, you are right. We already need to do something like that while deciding whether we should display the login form.